tests/hazmat/primitives/test_dh.py

```diff
@@ -102,3 +104,217 @@ def test_dh_public_numbers_equality():
+
+        serialized_params = params.parameters(backend)
+        serialized_public = public.public_key(backend)
+        serialized_private = private.private_key(backend)
+
+        assert isinstance(serialized_params, dh.DHParametersWithSerialization)
+        assert isinstance(serialized_public, dh.DHPublicKeyWithSerialization)
+        assert isinstance(serialized_private, dh.DHPrivateKeyWithSerialization)
+
+        params = dh.DHParameterNumbers(23, 18)
+        public = dh.DHPublicNumbers(1, params)
+        private = dh.DHPrivateNumbers(2, public)
+
+        with pytest.raises(ValueError):
+            private.private_key(backend)
+
+    def test_generate_dh(self, backend):
+        generator = 2
+        key_size = 512
+
+        parameters = dh.generate_parameters(generator, key_size, backend)
+        assert isinstance(parameters, dh.DHParameters)
+
+        key = parameters.generate_private_key()
+        assert isinstance(key, dh.DHPrivateKey)
+        assert key.key_size == key_size
+
+        public = key.public_key()
+        assert isinstance(public, dh.DHPublicKey)
+        assert public.key_size == key_size
+
+        assert isinstance(parameters, dh.DHParametersWithSerialization)
+        parameter_numbers = parameters.parameter_numbers()
+        assert isinstance(parameter_numbers, dh.DHParameterNumbers)
+        assert bit_length(parameter_numbers.p) == key_size
+
+        assert isinstance(public, dh.DHPublicKeyWithSerialization)
+        assert isinstance(public.public_numbers(), dh.DHPublicNumbers)
+        assert isinstance(public.parameters(), dh.DHParameters)
+
+        assert isinstance(key, dh.DHPrivateKeyWithSerialization)
+        assert isinstance(key.private_numbers(), dh.DHPrivateNumbers)
+        assert isinstance(key.parameters(), dh.DHParameters)
+
+    def test_tls_exchange(self, backend):
+        parameters = dh.generate_parameters(2, 512, backend)
+        assert isinstance(parameters, dh.DHParameters)
+
+        key1 = parameters.generate_private_key()
+        key2 = parameters.generate_private_key()
+
+        symkey1 = key1.exchange(key2.public_key())
+        assert symkey1
+        assert len(symkey1) == 512 // 8
+
+        symkey2 = key2.exchange(key1.public_key())
+        assert symkey1 == symkey2
+
+    def test_tls_exchange_algorithm(self, backend):
+        parameters = dh.generate_parameters(2, 512, backend)
+
+        key1 = parameters.generate_private_key()
+        key2 = parameters.generate_private_key()
+
+        shared_key_bytes = key2.exchange(key1.public_key())
+        symkey = int_from_bytes(shared_key_bytes, 'big')
+
+        symkey_manual = pow(key1.public_key().public_numbers().y,
+                            key2.private_numbers().x,
+                            parameters.parameter_numbers().p)
+
+        assert symkey == symkey_manual
+
+    def test_symmetric_key_padding(self, backend):
+        """
+        This test has specific parameters that produce a symmetric key
```

Comment:
I downloaded the zip to take a look and it appears none of it is useful for this.

I'm uncomfortable merging this without a source of vectors where we can load two keys, perform a handshake, and verify that the shared key is correct though...